src/storage/hummock_trace/src/collector.rs

```diff
@@ -0,0 +1,513 @@
+    LazyLock::new(|| UniqueIdGenerator::new(AtomicU64::new(0)));
+
+// Flag indicating whether trace should be used
+static SHOULD_USE_TRACE: LazyLock<bool> = LazyLock::new(set_should_use_trace);
+
+// Concurrent record ID generator for generating unique record IDs in concurrent environments
+pub static CONCURRENT_ID: LazyLock<ConcurrentIdGenerator> =
+    LazyLock::new(|| UniqueIdGenerator::new(AtomicU64::new(0)));
+
+// A tokio runtime for hummock tracing
+static TRACE_RT: LazyLock<Runtime> = LazyLock::new(|| Runtime::new().unwrap());
+
+pub const USE_TRACE: &str = "USE_HM_TRACE"; // Environment variable name for enabling trace
+const LOG_PATH: &str = "HM_TRACE_PATH"; // Environment variable name for specifying trace log path
+const DEFAULT_PATH: &str = ".trace/hummock.ht"; // Default trace log path
+const WRITER_BUFFER_SIZE: usize = 1024; // Buffer size for trace writer
+
+/// Returns whether trace should be used based on the environment variable
+pub fn should_use_trace() -> bool {
+    *SHOULD_USE_TRACE
+}
+
+/// Sets the value of the `SHOULD_USE_TRACE` flag based on the `USE_TRACE` environment variable
+fn set_should_use_trace() -> bool {
+    match std::env::var(USE_TRACE) {
+        Ok(v) => v.parse().unwrap_or(false),
+        Err(_) => false,
+    }
+}
+
+/// Initialize the `GLOBAL_COLLECTOR` with configured log file
+pub fn init_collector() {
+    TRACE_RT.spawn(async move {
+        let path = env::var(LOG_PATH).unwrap_or_else(|_| DEFAULT_PATH.to_string());
+        let path = Path::new(&path);
+        tracing::info!("Hummock Tracing log path {}", path.to_string_lossy());
+
+        if let Some(parent) = path.parent() {
+            if !parent.exists() {
+                create_dir_all(parent).unwrap();
+            }
+        }
+        let f = OpenOptions::new()
+            .write(true)
+            .truncate(true)
+            .create(true)
+            .open(path)
+            .expect("failed to open log file");
+        let writer = BufWriter::with_capacity(WRITER_BUFFER_SIZE, f);
+        let writer = TraceWriterImpl::try_new_bincode(writer).unwrap();
+        GlobalCollector::run(writer);
+    });
+}
+
+/// `GlobalCollector` collects traced hummock operations.
+/// It starts a collector thread and writer thread.
+struct GlobalCollector {
+    tx: Sender<RecordMsg>,
+    rx: Mutex<Option<Receiver<RecordMsg>>>,
+}
+
+impl GlobalCollector {
+    fn new() -> Self {
+        let (tx, rx) = channel();
+        Self {
+            tx,
+            rx: Mutex::new(Some(rx)),
+        }
+    }
+
+    fn run(mut writer: impl TraceWriter + Send + 'static) -> tokio::task::JoinHandle<()> {
+        tokio::task::spawn_blocking(move || {
+            let mut rx = GLOBAL_COLLECTOR.rx.lock().take().unwrap();
+            while let Some(Some(r)) = rx.blocking_recv() {
+                writer.write(r).expect("failed to write hummock trace");
+            }
+            writer.flush().expect("failed to flush hummock trace");
```

Comment:
As GLOBAL_COLLECTOR is static, I suppose currently we don't ensure writer.flush() will be called in Line 121?

Reply:
Why the `writer.flush()` will not be called when it is static?

Reply:
GlobalCollector::finish is only called in GlobalCollector::drop. And static GLOBAL_COLLECTOR calls drop on process exit? Besides, no coordination between process exit and the GlobalCollector::run thread exit.

Reply:
Oh, I get your point. You are right, it is not called. It should be fixed, though in my previous running it works as expected